src/serial.rs

```diff
@@ -83,63 +83,83 @@ pub struct Serial<USART, PINS> {
-    {
-        // NOTE(unsafe) This executes only during initialisation
-        let rcc = unsafe { &(*RCC::ptr()) };
-
-        /* Enable clock for USART */
-        rcc.apb2enr.modify(|_, w| w.usart1en().set_bit());
-
-        // Calculate correct baudrate divisor on the fly
-        let brr = clocks.pclk().0 / baud_rate.0;
-        usart.brr.write(|w| unsafe { w.bits(brr) });
-
-        /* Reset other registers to disable advanced USART features */
-        usart.cr2.reset();
-        usart.cr3.reset();
-
-        /* Enable transmission and receiving */
-        usart.cr1.modify(|_, w| unsafe { w.bits(0xD) });
-
-        Serial { usart, pins }
-    }
-
-    pub fn split(self) -> (Tx<USART1>, Rx<USART1>) {
-        (
-            Tx {
-                _usart: PhantomData,
-            },
-            Rx {
-                _usart: PhantomData,
-            },
-        )
-    }
-    pub fn release(self) -> (USART1, PINS) {
-        (self.usart, self.pins)
+macro_rules! usart {
+    ($($USART:ident: ($usart:ident, $usartXen:ident, $apbenr:ident),)+) => {
+        $(
+            use crate::stm32::$USART;
+            impl<PINS> Serial<$USART, PINS> {
+            pub fn $usart(usart: $USART, pins: PINS, baud_rate: Bps, clocks: Clocks) -> Self
+            where
+                PINS: Pins<$USART>,
+            {
+                // NOTE(unsafe) This executes only during initialisation
+                let rcc = unsafe { &(*stm32::RCC::ptr()) };
+
+                /* Enable clock for USART */
+                rcc.$apbenr.modify(|_, w| w.$usartXen().set_bit());
+
+                // Calculate correct baudrate divisor on the fly
+                let brr = clocks.pclk().0 / baud_rate.0;
+                usart.brr.write(|w| unsafe { w.bits(brr) });
+
+                /* Reset other registers to disable advanced USART features */
+                usart.cr2.reset();
+                usart.cr3.reset();
+
+                /* Enable transmission and receiving */
+                usart.cr1.modify(|_, w| unsafe { w.bits(0xD) });
+
+                Serial { usart, pins }
+            }
+                }
+        )+
     }
 }
 
 #[cfg(any(feature = "stm32f042", feature = "stm32f030"))]
-impl embedded_hal::serial::Read<u8> for Rx<USART1> {
+usart! {
+    USART1: (usart1, usart1en, apb2enr),
+}
+#[cfg(any(
+    feature = "stm32f042",
+    feature = "stm32f030x8",
+    feature = "stm32f030xc"
+))]
+usart! {
+    USART2: (usart2, usart2en, apb1enr),
+}
+#[cfg(any(feature = "stm32f030xc"))]
+usart! {
+    USART3: (usart3, usart3en, apb1enr),
+    USART4: (usart4, usart4en, apb1enr),
+    USART5: (usart5, usart5en, apb1enr),
+    // the usart6en bit is missing
+    // USART6: (usart6, usart6en, apb2enr),
+}
+
+// It's s needed for the impls, but rustc doesn't recognize that
+#[allow(dead_code)]
+type SerialRegisterBlock = stm32::usart1::RegisterBlock;
+
+impl<USART> embedded_hal::serial::Read<u8> for Rx<USART>
+where
+    USART: Deref<Target = SerialRegisterBlock>,
```

Comment:
This again feels kinda sketchy to me, as it is relying on all the USART peripherals continuing to use the same register block definition.

Reply:
If it's going to break, we're getting a compile error, so in that's ok. In the worst case you can do some macro based approach where you can separately implement the different register blocks.